[Spark-16579][SparkR] add install.spark function

junyangq opened the pull request:
## What changes were proposed in this pull request?

Add an install_spark function to the SparkR package. User can run `install_spark()` to install Spark to a local directory within R. 

Updates:

Several changes have been made:
- `install.spark()`
  - check existence of tar file in the cache folder, and download only if not found
  - trial priority of mirror_url look-up: user-provided -> preferred mirror site from apache website -> hardcoded backup option
  - use 2.0.0
- `sparkR.session()`
  - can install spark when not found in `SPARK_HOME`
## How was this patch tested?

Manual tests, running the check-cran.sh script added in #14173.

junyangq committed:
add install_spark

junyangq committed:
add doc for install_spark

SparkQA commented:
**[Test build #62518 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62518/consoleFull)** for PR 14258 at commit [`9d52d19`](https://github.com/apache/spark/commit/9d52d191acb5a55e9173897aaf381072590d9d01).

SparkQA commented:
**[Test build #62518 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62518/consoleFull)** for PR 14258 at commit [`9d52d19`](https://github.com/apache/spark/commit/9d52d191acb5a55e9173897aaf381072590d9d01).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62518/
Test FAILed.

junyangq committed:
changes to conform to R code style

SparkQA commented:
**[Test build #62519 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62519/consoleFull)** for PR 14258 at commit [`89efb04`](https://github.com/apache/spark/commit/89efb0438adee2cb95a4029d2779be5490b0a256).

SparkQA commented:
**[Test build #62519 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62519/consoleFull)** for PR 14258 at commit [`89efb04`](https://github.com/apache/spark/commit/89efb0438adee2cb95a4029d2779be5490b0a256).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62519/
Test PASSed.

shivaram commented:
cc @felixcheung @mengxr @sun-rui

shivaram commented:
Thanks @junyangq I'll take a look at this today. One question I had is about adding `install_spark` as a fallback option in `sparkR.session` if the jars are not found. Can we add that in this PR or do you want to do that in a different PR ?

shivaram commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,84 @@
+  packageName <- paste0(version, "-bin-hadoop", hadoop_version)
+  if (is.null(local_dir)) {
+    local_dir <- getOption("spark.install.dir",
+                           rappdirs::app_dir("spark"))$cache()
```
This will mean we need to add `rappdirs` as a dependency of SparkR ?

junyangq replied:
Yes. Perhaps we can do the implementation to avoid such dependency?

shivaram replied:
yeah it looks like we just want the cache dir from https://github.com/hadley/rappdirs/blob/6b42011053ec9db2068de3f93d3be0a9197b4043/R/cache.r#L42

junyangq replied:
Sure - will do that!

junyangq commented:
I can add that in this PR @shivaram

junyangq committed:
add install into sparkR.session if spark jar is not found

felixcheung commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,84 @@
+# Functions to install Spark in case the user directly downloads SparkR
+# from CRAN.
+
+#' Download and Install Spark to Local Directory
```
This might be a bit confusing? "If I have the SparkR package running why do I have to install Spark"?

junyangq replied:
Yeah it could be, so would "Spark Core" clear off some confusion?

felixcheung commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,84 @@
+install_spark <- function(hadoop_version = NULL, url = NULL, local_dir = NULL) {
+  version <- paste0("spark-", spark_version_default())
+  hadoop_version <- hadoop_version_default()
+  packageName <- paste0(version, "-bin-hadoop", hadoop_version)
```
do we need to support the "user-provided hadoop" favor of Spark, eg. "http://d3kbcqa49mib13.cloudfront.net/spark-1.6.2-bin-without-hadoop.tgz"

junyangq replied:
Good catch - will add that. Thanks for pointing it out.

felixcheung commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,84 @@
+#' @note install_spark since 2.1.0
+install_spark <- function(hadoop_version = NULL, url = NULL, local_dir = NULL) {
+  version <- paste0("spark-", spark_version_default())
+  hadoop_version <- hadoop_version_default()
```
aren't we overwriting here? shouldn't we only set hadoop_version if it is not passed in?

also, is there a way to do some rudimentary check on hadoop_version format?
alternatively since hadoop version don't change frequently and Spark only officially support a limited subset, would it be better if hadoop_version is a default vector of versions and we do [match.arg](https://stat.ethz.ch/R-manual/R-devel/library/base/html/match.arg.html) to check

junyangq replied:
Thanks, I'll fix the overwriting issue and rudimentary check will also be added :) For me I slightly prefer to refer to some external source (e.g. csv file). In case any change is needed, we only have to change the file without touching the source code, and it would be easier if we are going to support multiple versions later? 

about this @mengxr and I had some discussion previously. We thought about maintaining a table online containing available mirror sites so that we can make changes when necessary without touching the R package.

junyangq replied:
What do you think? @shivaram

shivaram replied:
I think we can do 3 things for the mirror url:
(a) have a spark mirror url as an argument to this function so that organizations can use a custom url if they want to use their Spark fork or their internal mirror etc.
(b) have a remote list of mirrors (through say apache) that we use if the argument is NULL
(c) fall back to some hardcoded mirror in the csv file in case the fetch in (b) fails

Regd using the CSV file, I think the biggest pain point is that it will require one more update on each release and we might forget to do this. So using some existing pointer like the latest spark version in the website might be better (not sure if we have something easy to parse ?)

felixcheung replied:
if you mean [this](https://github.com/apache/spark/pull/14258#discussion_r71426992) I wasn't sure csv was a good suggestion but hardcoding a list in .R would be problematic.

I think this bring us into the question/policy around compatibility - should be follow [semantic versioning](http://semver.org/)? like many other package managers, we could say we support "2.*" and that the Spark Core API interfacing with SparkR will need to be fully backward compatible.

junyangq replied:
I first made some change according to @shivaram 's suggestion, which SGTM. For (b), we should use a stable url, say http://www.apache.org/dyn/closer.lua ?. @mengxr mentioned offline that the other issue he worried about was whether apache is able to choose a fast mirror site via that. 

(c) If the fetch fails, we can resort to the csv file as the backup plan.

In terms of compatibility, do we only consider one Spark version right now?

felixcheung replied:
http://www.apache.org/dyn.closer.lua returns a HTML page, so http://www.apache.org/dyn/closer.cgi?as_json=1 is better.

felixcheung commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,84 @@
+}
+
+spark_version_default <- function() {
+  "2.0.0-preview"
```
perhaps better this is from the spark_download.csv too?

felixcheung commented on R/pkg/inst/extdata/spark_download.csv:
```diff
@@ -0,0 +1,2 @@
+"url","default"
+"http://apache.osuosl.org",TRUE
```
I wouldn't recommend hardcoding or relying on certain mirrors - apache mirrors are sort of dynamic - it's better to pull from 
http://www.apache.org/dyn/closer.cgi?as_json=1
like
http://www.apache.org/dyn/closer.lua/spark/spark-1.6.2/spark-1.6.2-bin-hadoop2.6.tgz

junyangq replied:
Combined with the previous comment: It turns out this would be similar to what sparklyr did. They provide a remote table as well as one shipped with the package that lists supported versions and their download addresses.

sun-rui replied:
I don't see the strong reason for a CSV file listing available mirrors. Because we can support apache dynamic mirrors, and install_spark() has a parameter allowing user to pass a vector of mirror sites.

junyangq replied:
Yeah, using apache dynamic mirror is great as long as the address does not change. The reason I was thinking of maintaining a (remote) csv file is that we are able to apply any change of the mirror sites without touching the R package...

junyangq committed:
message when SPARK_HOME is non-empty

junyangq committed:
change options of spark mirror url

SparkQA commented:
**[Test build #62660 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62660/consoleFull)** for PR 14258 at commit [`98087ad`](https://github.com/apache/spark/commit/98087ad0d5ae21322c7cef14200d093bcdf15c0e).

SparkQA commented:
**[Test build #62660 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62660/consoleFull)** for PR 14258 at commit [`98087ad`](https://github.com/apache/spark/commit/98087ad0d5ae21322c7cef14200d093bcdf15c0e).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62660/
Test FAILed.

junyangq committed:
minor changes

SparkQA commented:
**[Test build #62662 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62662/consoleFull)** for PR 14258 at commit [`0db89b7`](https://github.com/apache/spark/commit/0db89b7c98a42f4ca151f7ff07a4f49ee5ad87a6).

SparkQA commented:
**[Test build #62662 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62662/consoleFull)** for PR 14258 at commit [`0db89b7`](https://github.com/apache/spark/commit/0db89b7c98a42f4ca151f7ff07a4f49ee5ad87a6).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62662/
Test FAILed.

junyangq committed:
fix R style issue: don't use absolute paths

SparkQA commented:
**[Test build #62667 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62667/consoleFull)** for PR 14258 at commit [`503cb9f`](https://github.com/apache/spark/commit/503cb9fc2d4d54cf32defc835f568f30622e7925).

SparkQA commented:
**[Test build #62667 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62667/consoleFull)** for PR 14258 at commit [`503cb9f`](https://github.com/apache/spark/commit/503cb9fc2d4d54cf32defc835f568f30622e7925).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62667/
Test FAILed.

sun-rui commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,160 @@
+#' @note install_spark since 2.1.0
+install_spark <- function(hadoop_version = NULL, mirror_url = NULL,
+                          local_dir = NULL) {
+  version <- paste0("spark-", spark_version_default())
```
no need to create a function for spark version. just use "packageVersion("SparkR")"

junyangq replied:
Sounds good. Thanks :)

sun-rui commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,160 @@
+      mirror_url <- mirror_url_default()
+    }
+    # This is temporary, should be removed when released
+    version <- "spark-releases/spark-2.0.0-rc4-bin"
```
why this version is different from the above?

junyangq replied:
We did this temporarily only because the current `sparkR.session()` uses some methods that are not in the 2.0 preview version available from apache website, and so we try to download the rc4 version instead. We have to hardcoded some part of the url since it has different pattern from those for other versions.

junyangq committed:
remove spark version function, and use return value of packageVersion instead

junyangq committed:
remove unnecessary dir create

SparkQA commented:
**[Test build #62688 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62688/consoleFull)** for PR 14258 at commit [`78d6f91`](https://github.com/apache/spark/commit/78d6f910b93ffb58dea15361d020b64cba6ec541).

SparkQA commented:
**[Test build #62688 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62688/consoleFull)** for PR 14258 at commit [`78d6f91`](https://github.com/apache/spark/commit/78d6f910b93ffb58dea15361d020b64cba6ec541).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62688/
Test FAILed.

junyangq committed:
fix issue that dir.exists not available before 3.2.0

junyangq committed:
another dir.exists

SparkQA commented:
**[Test build #62860 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62860/consoleFull)** for PR 14258 at commit [`e4fe002`](https://github.com/apache/spark/commit/e4fe0023b8847fb24da9f733948ac2f13493457c).

SparkQA commented:
**[Test build #62860 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62860/consoleFull)** for PR 14258 at commit [`e4fe002`](https://github.com/apache/spark/commit/e4fe0023b8847fb24da9f733948ac2f13493457c).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62860/
Test FAILed.

junyangq committed:
temporary change of test

junyangq committed:
recover test changes

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+# Functions to install Spark in case the user directly downloads SparkR
+# from CRAN.
+
+#' Download and Install Spark Core to Local Directory
```
* `Spark Core` -> `Apache Spark`. This downloads the full distribution.
* `Local Directory` -> `a Local Directory`

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+
+#' Download and Install Spark Core to Local Directory
+#' 
+#' \code{install_spark} downloads and installs Spark to local directory if
```
Is `install_spark` consistent with SparkR naming convention? We only used underscores in wrapped SQL functions. Shall we use `install.spark`? cc: @shivaram

junyangq replied:
It's interesting to see two different opinions about object names in R. In the style guide of [Advanced R](http://adv-r.had.co.nz/Style.html), underscore (`_`) is encouraged, while in [Google's R style guide](https://google.github.io/styleguide/Rguide.xml#identifiers), they say underscore should be avoided and dots (`.`) is preferred.

junyangq replied:
For the sake of consistency, perhaps we can use `install.spark`?

mengxr replied:
This is just a style choice and we don't need to fight between "_" and ".". Our goal is to have consistent names in SparkR, which means following existing names.

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+#' Download and Install Spark Core to Local Directory
+#' 
+#' \code{install_spark} downloads and installs Spark to local directory if
+#' it is not found. The Spark version we use is 2.0.0 (preview). Users can
```
* `2.0.0 (preview)` => The Spark version is the same as the SparkR version. (Otherwise we need to update this doc for each Spark release.)

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+#' specify a desired Hadoop version, the remote site, and the directory where
+#' the package is installed locally.
+#'
+#' @param hadoop_version Version of Hadoop to install, 2.4, 2.6,
```
* `hadoop_version` => `hadoopVersion` (camelCase for params)

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+#' the package is installed locally.
+#'
+#' @param hadoop_version Version of Hadoop to install, 2.4, 2.6,
+#'        2.7 (default) and without
```
* I would put `without` as the default to match the default download on http://spark.apache.org/downloads.html.
* Do not mention other Hadoop versions as they might change. Use a `@seealso` link to redirect users to the download page for available Hadoop versions.

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+#'
+#' @param hadoop_version Version of Hadoop to install, 2.4, 2.6,
+#'        2.7 (default) and without
+#' @param mirror_url the base URL of the repositories to use
```
* `mirrorUrl`
* mention `the directory layout should follow Apache mirrors (http://www.apache.org/dyn/closer.lua/spark/)`

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+#' @param hadoop_version Version of Hadoop to install, 2.4, 2.6,
+#'        2.7 (default) and without
+#' @param mirror_url the base URL of the repositories to use
+#' @param local_dir local directory that Spark is installed to
```
* `localDir` (`cacheDir` might be more accurate)
* document default values for each OS

junyangq replied:
Could `cacheDir` be a little misleading if the user has already installed Spark, and in a non-cache directory?

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+#' @note install_spark since 2.1.0
+install_spark <- function(hadoop_version = NULL, mirror_url = NULL,
+                          local_dir = NULL) {
+  version <- paste0("spark-", packageVersion("SparkR"))
```
Does it mean that we can only publish SparkR with released Spark versions? Then how to make patched releases, say "2.0.0-1"? Can we overwrite an existing release on CRAN?

cc: @felixcheung

felixcheung replied:
that's an excellent point. we could have a fourth digital that we remove to match Spark version:
http://r-pkgs.had.co.nz/description.html#version
```
> numeric_version("1.9-10")
[1] ‘1.9.10’
> numeric_version("1.9.0-10")
[1] ‘1.9.0.10’
```
In that case we could patch the SparkR package

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+install_spark <- function(hadoop_version = NULL, mirror_url = NULL,
+                          local_dir = NULL) {
+  version <- paste0("spark-", packageVersion("SparkR"))
+  hadoop_version <- match.arg(hadoop_version, supported_versions_hadoop())
```
See my comment below on `supported_versions_hadoop`.

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+                        paste0(version, "-bin-without-hadoop"),
+                        paste0(version, "-bin-hadoop", hadoop_version))
+  if (is.null(local_dir)) {
+    local_dir <- getOption("spark.install.dir", spark_cache_path())
```
Where is `spark.install.dir` documented? If this is only used by SparkR, we should have a more specific config name.

junyangq replied:
After second thought, I guess perhaps we don't need to set global options since installation is just a one-time thing?

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+
+  packageLocalDir <- file.path(local_dir, packageName)
+
+
```
remove extra empty line

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+
+  # can use dir.exists(packageLocalDir) under R 3.2.0 or later
+  if (!is.na(file.info(packageLocalDir)$isdir)) {
+    fmt <- "Spark %s for Hadoop %s has been installed."
```
There should be an option to force re-install Spark in case the local directory or file is corrupted. And please mention that in the message.

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+                          })
+    if (fetchFail) {
+      message("Try the backup option.")
+      mirror_sites <- tryCatch(read.csv(mirror_url_csv()),
```
Let's skip the CSV solution in this PR and use Apache mirrors as the default.

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+  invisible(packageLocalDir)
+}
+
+mirror_url_default <- function() {
```
`default_mirror_url`

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+}
+
+supported_versions_hadoop <- function() {
+  c("2.7", "2.6", "2.4", "without")
```
I don't think we should hardcode Hadoop versions. They might change in the next release. And people might use patched version numbers for their own Hadoop releases, e.g. CDH. Shall we use "without" as the default and throw an error if the user requested Hadoop version is not available for download?

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+  c("2.7", "2.6", "2.4", "without")
+}
+
+spark_cache_path <- function() {
```
Are there references about the implementation here?

junyangq replied:
Yes, those actually refer to the implementation of rappdirs/R/cache.r. Should add reference here.

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+    winAppPath <- Sys.getenv("%LOCALAPPDATA%", unset = NA)
+    if (is.null(winAppPath)) {
+      msg <- paste("%LOCALAPPDATA% not found.",
+                    "Please define or enter an installation path in loc_dir.")
```
Where to enter the path?

junyangq replied:
Perhaps give specific message to let user restart?

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+    }
+  } else if (.Platform$OS.type == "unix") {
+    if (Sys.info()["sysname"] == "Darwin") {
+      path <- file.path(Sys.getenv("HOME"), "Library/Caches", "spark")
```
Need to check whether the folder exists and create the folder if it is not. I got this error in the first run

~~~
Error in normalizePath(path, mustWork = TRUE) : 
  path[1]="/Users/meng/Library/Caches/spark": No such file or directory
~~~

junyangq replied:
I changed to `mustWork = FALSE`. The check and create operations come later :)

mengxr commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,155 @@
+        "spark")
+    }
+  } else {
+    stop("Unknown OS")
```
Include the OS name in the error message.

mengxr commented on R/pkg/R/utils.R:
```diff
@@ -689,3 +689,7 @@ getSparkContext <- function() {
   sc
 }
+
+master_is_local <- function(master) {
```
`is_master_local`

mengxr commented on R/pkg/R/utils.R:
```diff
@@ -689,3 +689,7 @@ getSparkContext <- function() {
 }
+
+master_is_local <- function(master) {
+  grepl("^local(\\[[0-9\\*]*\\])?$", master, perl = TRUE)
```
`"^local(\\[[0-9\\*]+\\])?$"` (`+` instead of `*`)

felixcheung replied:
maybe
`^local(\\[([0-9]+|[\\*]*)\\])?$`
so it won't match `local[3*]`?

junyangq committed:
minor fix

SparkQA commented:
**[Test build #62874 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62874/consoleFull)** for PR 14258 at commit [`124110a`](https://github.com/apache/spark/commit/124110a66118e6e8fcfd248aea03e54e9c4f1481).

SparkQA commented:
**[Test build #62874 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62874/consoleFull)** for PR 14258 at commit [`124110a`](https://github.com/apache/spark/commit/124110a66118e6e8fcfd248aea03e54e9c4f1481).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62874/
Test FAILed.

junyangq committed:
fix docs

junyangq committed:
delete csv file option

junyangq committed:
rename functions of default mirror and check local master

junyangq committed:
fix mustWork issue in normalizePath, and create cache folder if not found

junyangq committed:
output OS name if not matched

junyangq committed:
more specific message and reference for spark_cache_path

junyangq committed:
more precise message under free build

SparkQA commented:
**[Test build #62884 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62884/consoleFull)** for PR 14258 at commit [`cf0f66d`](https://github.com/apache/spark/commit/cf0f66db23757d5267d4ac8a69733281b82d2bec).

SparkQA commented:
**[Test build #62884 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62884/consoleFull)** for PR 14258 at commit [`cf0f66d`](https://github.com/apache/spark/commit/cf0f66db23757d5267d4ac8a69733281b82d2bec).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62884/
Test PASSed.

junyangq committed:
fix typo on hadoop version message

SparkQA commented:
**[Test build #62885 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62885/consoleFull)** for PR 14258 at commit [`4f4a899`](https://github.com/apache/spark/commit/4f4a899f3955db7f0cc004fa71c89f0d873435c6).

SparkQA commented:
**[Test build #62885 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62885/consoleFull)** for PR 14258 at commit [`4f4a899`](https://github.com/apache/spark/commit/4f4a899f3955db7f0cc004fa71c89f0d873435c6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62885/
Test PASSed.

junyangq committed:
remove the setting of global spark.install.dir option

junyangq committed:
add overwrite option and adjust layout (#chars per line) of doc

mengxr commented:
@junyangq @shivaram We might need to think about how to test `install.spark`. Any ideas? It seems hard to me because the distribution tar file is not available for version like 2.1.0-SNAPSHOT.

junyangq committed:
remove hardcoded hadoop versions

junyangq committed:
edit doc of install.spark

felixcheung commented:
How about Scala version? I think Spark 2.0.0 is the first release that will have both Scala 2.10 and 2.11 support published

junyangq committed:
check installation of spark only when using ordinary R console

junyangq committed:
minor fix of overwrite message

junyangq committed:
minor change of messages

junyangq committed:
set env var SPARK_HOME after installation

junyangq committed:
minor change of doc

SparkQA commented:
**[Test build #62911 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62911/consoleFull)** for PR 14258 at commit [`fa94e3c`](https://github.com/apache/spark/commit/fa94e3cc99e93aea708a609733bbe9364b904efe).

SparkQA commented:
**[Test build #62911 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62911/consoleFull)** for PR 14258 at commit [`fa94e3c`](https://github.com/apache/spark/commit/fa94e3cc99e93aea708a609733bbe9364b904efe).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62911/
Test FAILed.

junyangq committed:
delete trailing space

SparkQA commented:
**[Test build #62912 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62912/consoleFull)** for PR 14258 at commit [`22f2f78`](https://github.com/apache/spark/commit/22f2f786bceeb599645c12210e3f49e66378ba6c).

SparkQA commented:
**[Test build #62912 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62912/consoleFull)** for PR 14258 at commit [`22f2f78`](https://github.com/apache/spark/commit/22f2f786bceeb599645c12210e3f49e66378ba6c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62912/
Test PASSed.

junyangq committed:
celebrate 2.0 release

junyangq committed:
add explanation of remote url structure, change default hadoop version

junyangq committed:
change default mirror

SparkQA commented:
**[Test build #62918 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62918/consoleFull)** for PR 14258 at commit [`e421c30`](https://github.com/apache/spark/commit/e421c30dd3dc3522467eaeacf82641bd108a02e8).

SparkQA commented:
**[Test build #62918 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62918/consoleFull)** for PR 14258 at commit [`e421c30`](https://github.com/apache/spark/commit/e421c30dd3dc3522467eaeacf82641bd108a02e8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62918/
Test PASSed.

junyangq committed:
add additional a step of searching for suggested url and only use default if not found

junyangq committed:
separate sub functions

junyangq committed:
fix typos in fun defs

junyangq committed:
fix mirror path and typo

junyangq committed:
fix message

junyangq committed:
fix message

junyangq committed:
fix typo

junyangq committed:
message update

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62962/
Test PASSed.

shivaram commented on R/pkg/R/sparkR.R:
```diff
@@ -365,6 +365,23 @@ sparkR.session <- function(
     }
     overrideEnvs(sparkConfigMap, paramMap)
   }
+  if (!grepl(".*shell\\.R$", Sys.getenv("R_PROFILE_USER"), perl = TRUE)) {
```
Could you add a comment on what we are checking above these `if` statements ?

junyangq replied:
Yes, that basically checks if you are working from the sparkR shell or RStudio, and do not download in the former case.

felixcheung replied:
There's a programmatic check in Spark JVM side you might want to do that instead of hard coding a file name.

junyangq replied:
@felixcheung Would you mind explaining a little more about that?

shivaram replied:
I think what @felixcheung meant was that in the Java side we already have a check called `isLocal`[1]. Thus we can just call that function using `callJMethod` instead of writing a new one

[1] https://github.com/apache/spark/blob/338a98d65c8efe0c41f39a8dddeab7040dcda125/core/src/main/scala/org/apache/spark/SparkContext.scala#L244

shivaram commented:
Thanks @junyangq - this is looking pretty good. It would be good to add a test for this (some discussion below).

@mengxr The R package version is still set as 2.0.0 in `DESCRIPTION` - so we should be able to test this right now. In fact it might be good to be able to force the Spark version to use in `install_spark` instead of using `packageVersion` just for testing.

The other thing I worry is if downloading 300 MB and un-tar might take too long to run on Jenkins. Also if we do add a test on jenkins, we should clean up the download, cache dirs etc.

@felixcheung I think the Scala version is dictated by the default builds we do for release. So for example `Starting version 2.0, Spark is built with Scala 2.11 by default` -- Users who want to use Scala 2.10 need to build their own tgz anyways and can point `install_spark` to the relevant files.

felixcheung commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,232 @@
+#' \code{without-hadoop}.
+#'
+#' @param hadoopVersion Version of Hadoop to install. Default is \code{"2.7"}. It can take other
+#'                      version number in the format of "int.int".
```
in R, there is no "int" so might not want to reference that.
how about `format of "x.y" where x and y are integer`?

junyangq replied:
Yeah, that's a good point. Thanks :)

felixcheung commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,232 @@
+  } else if (grepl("^[0-9]+\\.[0-9]+$", hadoopVersion, perl = TRUE)) {
+    paste0("hadoop", hadoopVersion)
+  } else {
+    hadoopVersion
```
shouldn't it fail here? I think we shouldn't let arbitrary string through?

junyangq replied:
I was thinking about the name of user patched versions, e.g. cdh4 - wondering if there is a way to tell such names from other invalid ones?

felixcheung replied:
There is usually a format?
`http://archive.cloudera.com/cdh5/cdh/5/spark-1.6.0-cdh5.8.0.tar.gz`

so `spark-2.0.0` + `-something`?

junyangq replied:
Are we trying to validate the `-something` part? Without attempting to download, we still don't know if the input version is good or not? I might miss something here.

shivaram replied:
FWIW I dont think its worth verifying the `-something` part just because it could be `cdh` or `hdp` etc. If the file is not found, we should throw an appropriate error which the user can use to understand what went wrong.

felixcheung commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,232 @@
+#'                   \item Win XP:
+#'                         \file{C:\\Documents and Settings\\<username>\\Local Settings\\Application
+#'                         Data\\spark\\spark\\Cache}
+#'                   \item Win Vista:
```
ah, I think we should spell out Windows. Also, this should be Vista and later?
(do we really want to list of different Windows version? why not use `%APPDATA% instead - then we have one explanation for all Windows versions?)

junyangq replied:
Yeah sounds good.

felixcheung commented on R/pkg/R/sparkR.R:
```diff
@@ -365,6 +365,23 @@ sparkR.session <- function(
+        fmt <- paste0("Spark not found in SPARK_HOME: %s.\n",
+                      "To search in the cache directory. ",
+                      "Installation will start if not found.")
+        msg <- sprintf(fmt, sparkHome)
```
can't we do something like:
```
fmt <- paste0("Spark not found in SPARK_HOME: ",
                         sparkHome,
                       " .\nTo search in the cache directory. ",
                       "Installation will start if not found.")
         message(msg)
```

felixcheung commented on R/pkg/R/sparkR.R:
```diff
@@ -365,6 +365,23 @@ sparkR.session <- function(
+        sparkHome <- packageLocalDir
+      } else {
+        fmt <- "Spark package is found in SPARK_HOME: %s"
+        msg <- sprintf(fmt, sparkHome)
```
similarly here...

felixcheung commented:
right, the release was not posted when I made the note. I see there's only scala 2.10 now.

felixcheung commented on R/pkg/R/utils.R:
```diff
@@ -689,3 +689,7 @@ getSparkContext <- function() {
 }
+
+is_master_local <- function(master) {
+  grepl("^local(\\[[0-9\\*]+\\])?$", master, perl = TRUE)
```
I think you didn't see my comment here: https://github.com/apache/spark/pull/14258#discussion_r72333901

junyangq replied:
Ah sorry, I missed that... I guess it shouldn't match `local[]` based on [master url](https://spark.apache.org/docs/latest/submitting-applications.html#master-urls) so perhaps `^local(\\[([0-9]+|\\*+)\\])?$`

junyangq committed:
concise message, improve doc for windows, fix regex match

SparkQA commented:
**[Test build #62982 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62982/consoleFull)** for PR 14258 at commit [`64756de`](https://github.com/apache/spark/commit/64756de02cae427be9b6795af3f40abbd30fcc06).

SparkQA commented:
**[Test build #62982 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62982/consoleFull)** for PR 14258 at commit [`64756de`](https://github.com/apache/spark/commit/64756de02cae427be9b6795af3f40abbd30fcc06).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62982/
Test PASSed.

shivaram commented:
@junyangq I just ran the CRAN checks locally and I see the problem you ran into in #14357 -- The problem is that if we try to run tests which depend on a Java-side change in master but not in 2.0.0 then they will fail. I think this shouldn't be a problem in the long run in the sense that we should match the SparkR and Spark versions closely. However for the first cut I'm fine with re-opening #14357 and say disabling some of the tests temporarily ?

junyangq commented:
@shivaram Would rebuild after cleaning solved the problem, as in #14357? So you mean disabling those tests in this PR first?

shivaram commented:
Not sure clean + rebuild will solve the problem here. The problem here is that we load the Spark 2.0.0 JARs using `install_spark` (i.e. that didn't have the fix in #14095) and we use R test code in the master branch which has the updated unit test. Or in other words need to use R code which doesn't have the test. (i.e. branch-2.0) and the master branch cannot be used with Spark 2.0.0 JARs. 

This may not be a big problem if we only enable CRAN checks on branch-2.0, but it seems like disabling the tests as in #14357 is an easy way to avoid confusion for now.

junyangq commented:
I see - since the local test uses the downloaded JARs, not the one built from master source files, but the test code used is from master - this causes the problem. I will first disable the tests and run CRAN checks.

junyangq committed:
Disable (temporarily) some test cases of describe and summary functions

This is because change of output from 2.0 to 2.1. The downloaded JARs are 2.0, while the test code in the master branch assumes new output format.

junyangq commented:
@mengxr @shivaram @felixcheung Thank you for the discussion and review of this PR. For compatibility of SparkR version and the JARs, perhaps it's better to send the PR and try to merge into branch 2.0 first, and come back to this later when the new version of Spark package is available for download.

felixcheung commented on R/pkg/R/install.R:
```diff
@@ -36,7 +36,7 @@
 #'
 #' @param hadoopVersion Version of Hadoop to install. Default is \code{"2.7"}. It can take other
-#'                      version number in the format of "int.int".
+#'                      version number in the format of "x.y" where x and y are integer.
```
"are integers"?

junyangq replied:
Yes, thanks!

felixcheung commented:
For my comment on https://github.com/apache/spark/pull/14258#discussion_r72683426

Like this: https://github.com/apache/spark/blob/fa4bc8ea8bab1277d1482da370dac79947cac719/core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala#L767

junyangq commented:
@felixcheung Sorry I still didn't get there. It seems that internally it checks via `args.primaryResource`. I was wondering if there is a good way to access to that in sparkR. Thanks!

shivaram commented:
@junyangq Is #14448 different from this PR or is it the same one on branch-2.0 ? I can just merge this into two branches, so we dont need a new PR I think

junyangq commented:
@shivaram There is only one additional minor change there.  The reason I opened #14448 on branch-2.0 is because we download the 2.0 jars, and there are some api changes from 2.0 to current master (e.g. `showString`), so I guess it would cause some problem if we use the master R code with 2.0 jars.

shivaram commented:
I see - So I was thinking that we could merge this into master as well as its not going to fail any tests or affect any users building SparkR from source -- I dont think we make any promises about the master branch to users. As long as the same code works in branch-2.0 then we can just backport this (if we do want a separate PR for branch-2.0 thats fine, but its just easier to keep all the code review on one PR)

@felixcheung @mengxr Any other comments on this ?

junyangq commented:
Sounds good to me. It doesn't fail tests except for the cran one if you delete `--no-test`.

felixcheung commented:
I think we should go ahead with this and get some usage from the community if we could as early as possible.
LGTM - we could see if we could improve on how to detect if running from shell later.

felixcheung commented on R/pkg/R/sparkR.R:
```diff
@@ -365,6 +365,23 @@ sparkR.session <- function(
   }
+  # do not download if it is run in the sparkR shell
+  if (!grepl(".*shell\\.R$", Sys.getenv("R_PROFILE_USER"), perl = TRUE)) {
+    if (!nzchar(master) || is_master_local(master)) {
```
shouldn't we also fail if master != local but SPARK_HOME is not defined or spark jar is not in SPARK_HOME?

felixcheung replied:
to clarify, i mean this check isn't restricted to local only, right?

junyangq replied:
Ah, since the install function is restricted to local for now, perhaps we only have to do the check for local master? If so, then I think it would be better to flip the order of `if` statements...

felixcheung commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,230 @@
+#'                  and force re-install Spark (in case the local directory or file is corrupted)
+#' @return \code{install.spark} returns the local directory where Spark is found or installed
+#' @rdname install.spark
+#' @name install.spark
```
add @aliases

felixcheung replied:
I think it's suppose to have the parameter types like this:
```
#' @aliases withColumnRenamed,SparkDataFrame,character,character-method
```

junyangq replied:
I thought it was the format for S4 methods? Do we want to make it S4?

shivaram replied:
I'd vote to leave this as a S3 method. Also I think the CRAN checks should show the missing aliases if there are any

junyangq replied:
I think it's just a normal function for now. @shivaram Did you mean that or the S3 method? Just to clarify...

felixcheung replied:
ah you are right - I've missed the fact that's it's S3. I think as S3 is fine.

junyangq replied:
sounds good, thanks!

felixcheung commented on R/pkg/R/install.R:
```diff
@@ -0,0 +1,230 @@
+
+  isFail <- tryCatch(download.file(packageRemotePath, packageLocalPath),
+                     error = function(e) {
+                       message(sprintf("Fetch failed from %s", mirrorUrl))
```
might want to `print(e)` as well so it could debug why the download throws exception

junyangq replied:
Good point. Thanks!

junyangq committed:
Send message of reset SPARK_HOME in installation

junyangq committed:
Specify path in jsonUrl and add alias to install function doc

SparkQA commented:
**[Test build #63455 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63455/consoleFull)** for PR 14258 at commit [`d84ba06`](https://github.com/apache/spark/commit/d84ba062b752f9987a1e8be28de5edd6c57c7a8b).

SparkQA commented:
**[Test build #63455 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63455/consoleFull)** for PR 14258 at commit [`d84ba06`](https://github.com/apache/spark/commit/d84ba062b752f9987a1e8be28de5edd6c57c7a8b).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63455/
Test FAILed.

junyangq committed:
remove comment

SparkQA commented:
**[Test build #63456 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63456/consoleFull)** for PR 14258 at commit [`3aeb4eb`](https://github.com/apache/spark/commit/3aeb4ebe7b18f5b26914cdf730cced3eb536f48d).

SparkQA commented:
**[Test build #63456 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63456/consoleFull)** for PR 14258 at commit [`3aeb4eb`](https://github.com/apache/spark/commit/3aeb4ebe7b18f5b26914cdf730cced3eb536f48d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63456/
Test PASSed.

felixcheung commented:
LGTM

shivaram commented:
Thanks @junyangq and @felixcheung -- LGTM. Merging this to master and branch-2.0
We should add some tests to this and enable the checks to run on every PR. But we can do this as a part of SPARK-16577
